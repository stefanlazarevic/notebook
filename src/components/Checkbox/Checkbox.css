.Checkbox {
    box-sizing: border-box;

    position: relative;

    display: inline-block;

    cursor: pointer;

    width: 20px;

    height: 20px;

    font-size: 20px;

    background-color: #f8f9fb;

    border: 1px solid #e4e4e4;

    user-select: none;
}

.Checkbox:focus {
    outline: 2px solid #5149b8;

    outline-offset: -1px;
}

.Checkbox::before {
    content: '✓';

    box-sizing: border-box;

    position: absolute;

    display: none;

    width: 100%;

    height: 100%;

    top: 1px;

    left: 0;

    text-align: center;

    line-height: 0.9em;
}

.Checkbox[aria-checked="mixed"]::before {
    content: '━';

    color: #333;

    display: inline-block;

    font-size: 10px;
    
    line-height: 1rem;
}

.Checkbox[aria-checked="true"] {
    background-color: #7e6ee6;

    border-color: #7e6ee6;

    color: #fff;
}

.Checkbox[aria-checked="true"]::before {
    display: inline-block;
}

.Checkbox[aria-disabled="true"] {
    cursor: not-allowed;
}