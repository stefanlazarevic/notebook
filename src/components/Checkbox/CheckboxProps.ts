import PropTypes, { InferProps } from "prop-types";
import { Validator, AriaAttributes } from "react";

export const CheckboxPropTypes = {
	/**
	 * Јединствени идентификатор.
	 */
	id: PropTypes.string as Validator<string | undefined>,
	/**
	 * Јединствени идентификатор намењен за Unit и E2E тестове.
	 */
	testid: PropTypes.string as Validator<string | undefined>,
	/**
	 * Вредност овог атрибута дефинише стање поља за потврду.
	 * Вредности могу бити `true`, `false` и "mixed".
	 *
	 * "mixed" вредност представља неодређено стање.
	 *
	 * @default false
	 */
	checked: PropTypes.oneOf([true, false, "true", "false", "mixed"]) as Validator<AriaAttributes["aria-checked"]>,
	/**
	 * Назив ознаке која описује поље за потврду.
	 */
	"aria-labelledby": PropTypes.string as Validator<string | undefined>,
	/**
	 *
	 */
	"aria-label": PropTypes.string as Validator<string | undefined>,
	/**
	 *
	 */
	"aria-describedby": PropTypes.string as Validator<string | undefined>,
	/**
	 *
	 */
	"aria-haspopup": PropTypes.oneOf([true, false, "true", "false"]) as Validator<boolean | "false" | "true" | undefined>,
	/**
	 * Повратни позив који се извршава када дође до промене стања поља.
	 */
	onChange: PropTypes.func as Validator<
		(
			event: React.SyntheticEvent,
			currentCheckedState: boolean | "false" | "mixed" | "true" | undefined,
			index?: number
		) => void | undefined
	>,
	/**
	 * Повратни позив који се извршава приликом употребе `Shift` + `F10` пречице.
	 */
	onContext: PropTypes.func as Validator<(event: React.SyntheticEvent) => void | undefined>,
	/**
	 * Називи CSS класа.
	 */
	className: PropTypes.string as Validator<string | undefined>,
	/**
	 * Индикатор онемогућености поља.
	 */
	disabled: PropTypes.bool as Validator<boolean | undefined>,
	/**
	 * Информација о позицији поља у колекцији.
	 * Уколико је вредност дефинисана, приликом промене стања поља, ова информација
	 * се прослеђује као трећи параметар.
	 */
	index: PropTypes.number as Validator<number | undefined>,
};

export type CheckboxProps = InferProps<typeof CheckboxPropTypes>;
