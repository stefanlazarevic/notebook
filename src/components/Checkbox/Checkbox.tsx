import React from "react";

import "./Checkbox.css";

import { CheckboxProps, CheckboxPropTypes } from "./CheckboxProps";

import useClassNames from "../Utils/hooks/classNames";

function Checkbox(props: CheckboxProps) {
	const className = useClassNames("Checkbox", props.className);

	/**
	 * Функција која приликом промене стања извршава повратни позив уколико је
	 * дефинисан.
	 *
	 * @param event Акција извршена путем миша или тастатуре.
	 */
	function change(event: React.SyntheticEvent) {
		if (typeof props.onChange === "function" && !props.disabled) {
			props.onChange(event, props.checked, props.index);
		}
	}

	/**
	 * ...
	 * @param event Акција извршена путем миша.
	 */
	function onClick(event: React.MouseEvent) {
		change(event);
	}

	/**
	 * ...
	 * @param event Акција извршена путем тастатуре.
	 */
	function onKeyDown(event: React.KeyboardEvent) {
		const { keyCode, shiftKey } = event;

		if (keyCode === 32) {
			change(event);
		}

		if (shiftKey && keyCode === 121 && typeof props.onContext === "function") {
			props.onContext(event);
		}
	}

	return (
		<div
			id={props.id}
			data-testid={props.testid}
			data-index={props.index}
			tabIndex={props.disabled ? undefined : 0}
			role="checkbox"
			className={className}
			aria-checked={props.checked}
			aria-labelledby={props["aria-labelledby"]}
			aria-label={props["aria-label"]}
			aria-describedby={props["aria-describedby"]}
			aria-haspopup={props["aria-haspopup"]}
			aria-disabled={props.disabled}
			onClick={props.disabled ? undefined : onClick}
			onKeyDown={props.disabled ? undefined : onKeyDown}
		/>
	);
}

Checkbox.propTypes = CheckboxPropTypes;

Checkbox.defaultProps = {
	onChange: undefined,
	onContext: undefined,
};

Checkbox.displayName = "Checkbox";

export default Checkbox;
