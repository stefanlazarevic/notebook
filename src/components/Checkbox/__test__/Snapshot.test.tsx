import React from "react";
import { render, cleanup } from "@testing-library/react";
import Checkbox from "../index";

afterEach(cleanup);

it("Подразумеванa својства се подударају са последњим снимком.", () => {
	const { asFragment } = render(<Checkbox onChange={() => void 0} />);

	expect(asFragment()).toMatchSnapshot();
});

it("Сва подешена својства се подударају са последњим снимком.", () => {
	const { asFragment } = render(
		<Checkbox
			id="id"
			testid="test-id"
			aria-labelledby="test-label"
			checked={true}
			onChange={() => void 0}
			className="CustomCheckbox"
			disabled={false}
		/>
	);

	expect(asFragment()).toMatchSnapshot();
});
