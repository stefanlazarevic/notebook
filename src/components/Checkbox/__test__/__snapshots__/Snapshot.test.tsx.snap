// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Подразумеванa својства се подударају са последњим снимком. 1`] = `
<DocumentFragment>
  <div
    class="Checkbox"
    role="checkbox"
    tabindex="0"
  />
</DocumentFragment>
`;

exports[`Сва подешена својства се подударају са последњим снимком. 1`] = `
<DocumentFragment>
  <div
    aria-checked="true"
    aria-disabled="false"
    aria-labelledby="test-label"
    class="Checkbox CustomCheckbox"
    data-testid="test-id"
    id="id"
    role="checkbox"
    tabindex="0"
  />
</DocumentFragment>
`;
